core-bundle/src/DataContainer/OperationsFactory.php

```diff
@@ -0,0 +1,150 @@
+{
+    public function __construct(private Security $security)
+    {
+    }
+
+    public function getForTable(string $table): array
+    {
+        $defaults = $this->getDefaults($table);
+        $dca = $GLOBALS['TL_DCA'][$table]['list']['operations'] ?? null;
+
+        if (!\is_array($dca)) {
+            return $defaults;
+        }
+
+        $operations = [];
+
+        // If none of the defined operations are name-only, we append the operations to the defaults.
+        if (empty(array_filter($dca, static fn($v) => \is_string($v) && isset($defaults[$v])))) {
+            $operations = $defaults;
+        }
+
+        foreach ($dca as $k => $v) {
+            if (\is_string($v) && isset($defaults[$v])) {
+                $operations[$v] = $defaults[$v];
+                continue;
+            }
+
+            $v = \is_array($v) ? $v : [$v];
+
+            // If the operation exists but only has access_callback, the callbacks were probably added
+            // by an event listener and should just be merged with the default config.
+            if (1 === \count($v) && \is_array($v['access_callback'] ?? null)) {
+                $callbacks = $v['access_callback'];
+                $v = $defaults[$k];
+                $v['access_callback'] = array_merge($callbacks, $v['access_callback'] ?? []);
+            }
+
+            $operations[$k] = $v;
+        }
+
+        return $operations;
+    }
+
+    private function getDefaults(string $table): array
+    {
+        $operations = [];
+
+        $isTreeMode = ($GLOBALS['TL_DCA'][$table]['list']['sorting']['mode'] ?? null) === DataContainer::MODE_TREE;
+        $hasPtable = !empty($GLOBALS['TL_DCA'][$table]['config']['ptable'] ?? null);
+        $ctable = $GLOBALS['TL_DCA'][$table]['config']['ctable'][0] ?? null;
+
+        if ($ctable) {
+            $operations += [
+                'edit' => [
+                    'href' => 'table='.$ctable,
+                    'icon' => 'edit.svg',
+                ],
+                'editheader' => [
+                    'href' => 'act=edit',
+                    'icon' => 'header.svg',
+                    'access_callback' => [$this->accessCallback(ContaoCorePermissions::DC_ACTION_EDIT)],
+                ],
+            ];
+        } else {
+            $operations['edit'] = [
+                'href' => 'act=edit',
+                'icon' => 'edit.svg',
+                'access_callback' => [$this->accessCallback(ContaoCorePermissions::DC_ACTION_EDIT)],
+            ];
+        }
+
+        if ($hasPtable || $isTreeMode) {
+            $operations['copy'] = [
+                'href' => 'act=paste&amp;mode=copy',
+                'icon' => 'copy.svg',
+                'attributes' => 'onclick="Backend.getScrollOffset()"',
+                'access_callback' => [$this->accessCallback(ContaoCorePermissions::DC_ACTION_COPY)],
+            ];
+
+            if ($isTreeMode) {
+                // TODO: how to we check permissions for that?
```

Comment:
Guess it needs to be a Core permission because it's a core action?

Reply:
Nah, I just need to figure out common grounds. `tl_page` checks if there are child records on the page, so maybe thats what the "permission check" should do. As well as the usual "copy" checks. But since I only focussed on `tl_news_archive` you provided this is not finished yet 🙃